src/arithmetic/path_funcs/path_funcs.c

```diff
@@ -0,0 +1,179 @@
+	/* Reverse direction if needed. A direction change is needed if the last node in the path, reading
+	 * RTL is the source node in the edge, and the edge direction in the query is LTR.
+	 * path =[(a)]
+	 * e = (a)->(b)
+	 * Query: MATCH p=(a)<-[]-(b)
+	 * e direction needs to be change. */
+
+	if(RTLEdge && nId == edge->srcNodeID) edge = _Edge_ReverseDirection(edge);
+	Path_AppendEdge(path, *edge);
+}
+
+static void _PathBuilder_AppendPath(Path *path, Path *new_path, bool RTLEdge) {
+	uint path_node_count = Path_NodeCount(*path);
+	assert(path_node_count > 0);
+	// No need to append empty paths.
+	uint new_path_node_count = Path_NodeCount(*new_path);
+	if(new_path_node_count == 0) return;
+
+	Node *last_LTR_node = &path->nodes[path_node_count - 1];
+	EntityID last_LTR_node_id = ENTITY_GET_ID(last_LTR_node);
+	Node *new_path_node_0 = &new_path->nodes[0];
+	EntityID new_path_node_0_id = ENTITY_GET_ID(new_path_node_0);
+	Node *new_path_last_node = &new_path->nodes[new_path_node_count - 1];
+	EntityID new_path_last_node_id = ENTITY_GET_ID(new_path_last_node);
+	// Validate current last LTR node is in either edges of the path.
+	assert(last_LTR_node_id == new_path_node_0_id || last_LTR_node_id == new_path_last_node_id);
+	int new_path_edge_count = Path_EdgeCount(*new_path);
+	// Check if path needs to be rverse inserated or not.
+	if(last_LTR_node_id == new_path_last_node_id) {
+		// Insert only nodes which are not the last and the first, since they will be added by append node specifically.
+		_PathBuilder_AppendEdge(path, &new_path->edges[new_path_edge_count - 1], RTLEdge);
+		for(int i = new_path_edge_count - 2; i >= 0; i++) {
+			_PathBuilder_AppendNode(path, &new_path->nodes[i + 1]);
+			_PathBuilder_AppendEdge(path, &new_path->edges[i], RTLEdge);
+		}
+	} else {
+		// Normal insertion.
+		for(uint i = 0; i < new_path_edge_count - 1; i++) {
+			// Insert only nodes which are not the last and the first, since they will be added by append node specifically.
+			_PathBuilder_AppendNode(path, &new_path->nodes[i + 1]);
+			_PathBuilder_AppendEdge(path, &new_path->edges[i], RTLEdge);
+		}
+		_PathBuilder_AppendEdge(path, &new_path->edges[new_path_edge_count - 1], RTLEdge);
+	}
+}
+
+/* Creates a path from a given sequence of graph entities.
+ * The first argument is the ast represents the path.
+ * Arguments 2...n are the sequence of graph entities combines the path.
+ * The sequence is always in odd length and defined as:
+ * Odd indices members are always representing the value of a single node.
+ * Even indices members are either representing the value of a single edge,
+ * or an intermidate sipath, in case of variable length traversal. */
+SIValue AR_TOPATH(SIValue *argv, int argc) {
+	const cypher_astnode_t *ast_path = argv[0].ptrval;
+	uint nelements = cypher_ast_pattern_path_nelements(ast_path);
+	assert(argc == (nelements + 1));
+	Path *path = rm_malloc(sizeof(Path));
+	*path = Path_New(nelements);
```

Comment:
This is a bit strange I wouldn't expect 2 calls which seems to be initialised the object.

Reply:
changed